demo/lightning/checkpoint/multinode/train.py

```diff
@@ -0,0 +1,237 @@
+import io
+import os
+import socket
+import time
+import torch
+
+from contextlib import contextmanager
+from typing import Generator
+from lightning import Trainer
+from lightning.pytorch.callbacks import ModelCheckpoint
+from lightning.pytorch.demos import (LightningTransformer, Transformer,
+                                     WikiText2)
+from torch.utils.data import DataLoader
+from pathlib import Path
+from google.cloud import storage
+
+from dataflux_pytorch.lightning.gcs_filesystem import GCSDistributedWriter, GCSDistributedReader
+from dataflux_pytorch.lightning.path_utils import parse_gcs_path
+from dataflux_core import user_agent
+from dataflux_pytorch.lightning import DatafluxLightningCheckpoint
+from lightning.pytorch.trainer.states import TrainerFn
+from lightning.pytorch.strategies import FSDPStrategy
+from lightning.pytorch.strategies.fsdp import _METADATA_FILENAME
+from torch.distributed.checkpoint import save, load
+from torch.nn import Module
+
+
+class DatafluxFSDPStrategy(FSDPStrategy):
+
+    def __init__(self, path, project_name, storage_client, model, **kwargs):
+        super().__init__(**kwargs)
+        self.writer = GCSDistributedWriter(path, project_name, storage_client)
+        self.reader = GCSDistributedReader(
+            path, project_name, storage_client)
+        self.checkpoint_io = DatafluxLightningCheckpoint(
+            project_name, storage_client)
+        self.model = model
+        self.storage_client = storage.Client(project=project_name,)
+        user_agent.add_dataflux_user_agent(self.storage_client)
+
+    def save_checkpoint(self,
+                        checkpoint,
+                        filepath,
+                        storage_options=None) -> None:
+        if storage_options is not None:
+            raise TypeError(
+                "`FSDPStrategy.save_checkpoint(..., storage_options=...)` is not supported because"
+                " `FSDPStrategy` does not use the `CheckpointIO`.")
+
+        path = Path(self.broadcast(filepath))
+
+        converted_state = {"model": checkpoint.pop("state_dict")}
+        converted_state.update({
+            f"optimizer_{idx}": optim_state
+            for idx, optim_state in enumerate(
+                checkpoint.pop("optimizer_states", []))
+        })
+        save(converted_state, checkpoint_id=path, storage_writer=self.writer)
+
+        if self.global_rank == 0:
+            self.checkpoint_io.save_checkpoint(checkpoint,
+                                               path / _METADATA_FILENAME)
+
+    def get_sharded_state_dict_context(self, module: Module) -> Generator[None, None, None]:
+        from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
```

Comment:
Is there a specific reason this is imported within the function? since there is a second import of this on line 83 should we just import this in the global scope?

Reply:
Hmm that makes sense. This is sort of copied over from FSDP load_checkpoint implementation and hence I didnt refactor it. 

Moved to global import.